common/splice_script.c

```diff
@@ -0,0 +1,2696 @@
+						struct splice_script_result ***result)
+{
+	struct splice_script_error *error;
+	struct token **tokens = tal_arr(ctx, struct token *,
+					strlen(script) + 1);
+
+	for (size_t i = 0; i < strlen(script); i++) {
+		tokens[i] = new_token(tokens, is_delimiter(script[i])
+						? TOK_DELIMITER
+						: TOK_CHAR, i);
+		tokens[i]->c = script[i];
+	}
+
+	/* We add a delimiter on the end to make life simple. */
+	tokens[strlen(script)] = new_token(tokens, TOK_DELIMITER,
+					   strlen(script));
+
+	if ((error = clean_whitespace(ctx, &tokens)))
+		return error;
+
+	if ((error = find_arrows_and_strs(ctx, &tokens)))
+		return error;
+
+	if ((error = process_top_separators(ctx, &tokens)))
+		return error;
+
+	if ((error = process_2nd_separators(ctx, &tokens)))
+		return error;
+
+	if ((error = process_3rd_separators(ctx, &tokens)))
+		return error;
+
+	if ((error = data_types(ctx, channels, channels_count, &tokens)))
+		return error;
+
+	if ((error = compress_top_operands(ctx, &tokens)))
+		return error;
+
+	if ((error = compress_2nd_operands(ctx, &tokens)))
+		return error;
+
+	if ((error = resolve_channel_ids(ctx, channels, channels_count,
+					 &tokens)))
+		return error;
+
+	if ((error = make_segments(ctx, &tokens)))
+		return error;
+
+	if ((error = expand_multichans(ctx, &tokens)))
+		return error;
+
+	if ((error = validate_and_clean(ctx, channels, channels_count,
+					&tokens)))
+		return error;
+
+	if ((error = calculate_amounts(ctx, &tokens)))
+		return error;
+
+	/* Internal token diagnostics:
+	return debug_dump(ctx, tokens);
+	return dump_segments(ctx, tokens);
+	*/
+
+	(void)debug_dump;
+	(void)dump_segments;
+
+	*result = tal_arr(ctx, struct splice_script_result*, tal_count(tokens));
+
+	for (size_t i = 0; i < tal_count(tokens); i++) {
+		(*result)[i] = talz(*result, struct splice_script_result);
+		struct splice_script_result *itr = (*result)[i];
+		struct token *lease = tokens[i]->left->right;
+		struct token *fee = tokens[i]->left->middle
+					?: tokens[i]->right->middle;
+
+		if (lease) {
+			itr->lease_sat = lease->amount_sat;
+			itr->lease_max_ppm = lease->ppm;
+		}
+
+		itr->in_sat = tokens[i]->left->amount_sat;
+		itr->in_ppm = tokens[i]->left->ppm;
+
+		if (tokens[i]->middle->type == TOK_CHANID)
+			itr->channel_id = tal_dup(*result, struct channel_id,
+						  tokens[i]->middle->chan_id);
+		else if (tokens[i]->middle->type == TOK_BTCADDR)
+			itr->bitcoin_address = tal_strdup(*result,
+							 tokens[i]->middle->str);
+		else if (tokens[i]->middle->type == TOK_WALLET)
+			itr->onchain_wallet = true;
+		else
+			return new_error(ctx, INTERNAL_ERROR, tokens[i],
+					 "splice_script_result");
+
+		itr->out_sat = tokens[i]->right->amount_sat;
+		itr->out_ppm = tokens[i]->right->ppm;
+
+		if (tokens[i]->right->type == TOK_WILDCARD)
+			itr->out_ppm = UINT32_MAX;
+
+		if (fee) {
+			itr->pays_fee = true;
+			itr->feerate_per_kw = fee->amount_sat.satoshis; /* Raw: We use satoshis as our base number unit */
```

Comment:
little weird?

Reply:
Duplicate of https://github.com/ElementsProject/lightning/pull/6980#discussion_r1569444636